### PR TITLE
Code improvements

### PR DESCRIPTION
**Pull request by ingvagabund**
- simplifying the code
- making the code more readable

### CODE REVIEW

File: pkg/e2e/autoscaler/autoscaler.go [hunk 1]
```diff
@@ -207,40 +218,24 @@ var _ = g.Describe("[Feature:Machines] Autoscaler should", func() {
...
-			wait.PollImmediate(1*time.Second, e2e.WaitShort, func() (bool, error) {
-				if err := client.Delete(context.TODO(), machineAutoscaler); err != nil {
-					glog.Errorf("error querying api for machineAutoscaler object: %v, retrying...", err)
-					return false, nil
-				}
-				return true, nil
-			})
+			err = e2e.DeleteObjectsByLabels(context.TODO(), client, map[string]string{autoscalingTestLabel: ""}, &caov1alpha1.MachineAutoscalerList{})
+			if err != nil {
+				glog.Warning(err)
```
Comment (frobware) [inline pkg/e2e/autoscaler/autoscaler.go]: It is OK to continue in the face of an error?

File: pkg/e2e/framework/common.go [hunk 2]
```diff
@@ -9,3 +20,43 @@ const (
 	WaitLong        = 10 * time.Minute
 	RetryMedium     = 5 * time.Second
 )
+
+// DeleteObjectsByLabels list all objects of a given kind by labels and deletes them.
+// Currently supported kinds:
+// -
```
Comment (frobware) [inline pkg/e2e/framework/common.go]: To be continued... ?

File: pkg/e2e/autoscaler/autoscaler.go [hunk 3]
```diff
@@ -197,25 +196,16 @@ var _ = g.Describe("[Feature:Machines] Autoscaler should", func() {
...
 		}
 
-		g.By(fmt.Sprintf("Create ClusterAutoscaler and MachineAutoscaler objects. Targeting machineSet %s", targetMachineSet.Name))
-		initialNumberOfReplicas := *targetMachineSet.Spec.Replicas
+		g.By("Create ClusterAutoscaler object")
 		clusterAutoscaler := clusterAutoscalerResource()
+		err = client.Create(context.TODO(), clusterAutoscaler)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		initialNumberOfReplicas := *targetMachineSet.Spec.Replicas
```
Comment (frobware) [inline pkg/e2e/autoscaler/autoscaler.go]: This is a general comment for the PR: we should choose to deref safely using `pointer.Int32PtrDerefOr()`.

File: pkg/e2e/autoscaler/autoscaler.go [hunk 4]
```diff
@@ -117,175 +121,151 @@ func labelMachineSetNodes(client runtimeclient.Client, ms *mapiv1beta1.MachineSe
 	})
 }
 
+// Build default CA resource to allow fast scaling up and down
+func clusterAutoscalerResource() *caov1alpha1.ClusterAutoscaler {
```
Comment (enxebre) [inline pkg/e2e/autoscaler/autoscaler.go]: can we move this into utils?

File: pkg/e2e/autoscaler/autoscaler.go [hunk 5]
```diff
@@ -117,175 +121,151 @@ func labelMachineSetNodes(client runtimeclient.Client, ms *mapiv1beta1.MachineSe
...
-			}
-			return true, nil
-		})
+		g.By("Create ClusterAutoscaler object")
+		clusterAutoscaler := clusterAutoscalerResource()
+		err = client.Create(context.TODO(), clusterAutoscaler)
 		o.Expect(err).NotTo(o.HaveOccurred())
 
-		workLoad := newWorkLoad()
+		initialNumberOfReplicas := *targetMachineSet.Spec.Replicas
```
Comment (enxebre) [inline pkg/e2e/autoscaler/autoscaler.go]: can we use `pointer.Int32PtrDerefOr`

File: pkg/e2e/autoscaler/autoscaler.go [hunk 6]
```diff
@@ -117,175 +121,151 @@ func labelMachineSetNodes(client runtimeclient.Client, ms *mapiv1beta1.MachineSe
...
+		g.By("Waiting for cluster to scale out number of replicas")
 		err = wait.PollImmediate(5*time.Second, e2e.WaitLong, func() (bool, error) {
-			msKey := types.NamespacedName{
-				Namespace: e2e.TestContext.MachineApiNamespace,
-				Name:      targetMachineSet.Name,
-			}
 			ms := &mapiv1beta1.MachineSet{}
-			if err := client.Get(context.TODO(), msKey, ms); err != nil {
-				glog.Errorf("error querying api for clusterAutoscaler object: %v, retrying...", err)
+			if err := client.Get(context.TODO(), targetMachineSetKey, ms); err != nil {
```
Comment (enxebre) [inline pkg/e2e/autoscaler/autoscaler.go]: can we create `getMachineSet` in utils?

File: pkg/e2e/infra/infra.go [hunk 7]
```diff
@@ -34,10 +35,15 @@ var _ = g.Describe("[Feature:Machines] Managed cluster should", func() {
 		client, err := e2e.LoadClient()
 		o.Expect(err).NotTo(o.HaveOccurred())
 
-		machines, err := getMachines(client)
+		machineList := mapiv1beta1.MachineList{}
```
Comment (enxebre) [inline pkg/e2e/infra/infra.go]: can we update getMachines to use `client.List` and keep this as it was?

File: pkg/e2e/autoscaler/autoscaler.go (hunk 4 above)
Comment (ingvagabund) [reply pkg/e2e/autoscaler/autoscaler.go]: I prefer to keep it in the same file as it is autoscaling specific. Unless it is needed in a different `e2e/PACKAGE` suite. The less we expose/export the better.

File: pkg/e2e/autoscaler/autoscaler.go (hunk 5 above)
Comment (ingvagabund) [reply pkg/e2e/autoscaler/autoscaler.go]: Wondering if that would help. I would rather panic in case `targetMachineSet.Spec.Replicas == nil` rather than set some default value and then debug unexpected behaviour caused by it for 1h.

File: pkg/e2e/autoscaler/autoscaler.go (hunk 6 above)
Comment (ingvagabund) [reply pkg/e2e/autoscaler/autoscaler.go]: I removed the function on purpose since it can be replaced by the one-liner you are commenting onto.

File: pkg/e2e/infra/infra.go (hunk 7 above)
Comment (ingvagabund) [reply pkg/e2e/infra/infra.go]: I removed `getMachines` function on purpose. As long as it can be replaced by a single invocation, we should not hide it in another function/method.

File: pkg/e2e/autoscaler/autoscaler.go (hunk 1 above)
Comment (ingvagabund) [reply pkg/e2e/autoscaler/autoscaler.go]: It's in the defer part where all the created resources are deleted. So, if this one fails, there are still other resources to be deleted.

File: pkg/e2e/framework/common.go (hunk 2 above)
Comment (ingvagabund) [reply pkg/e2e/framework/common.go]: Yes, I will fill it up.

File: pkg/e2e/autoscaler/autoscaler.go (hunk 1 above)
Comment (frobware) [reply pkg/e2e/autoscaler/autoscaler.go]: Maybe good to add:

```Go
// if this one fails, there are still other resources to be deleted.
```

File: pkg/e2e/autoscaler/autoscaler.go (hunk 5 above)
Comment (ingvagabund) [reply pkg/e2e/autoscaler/autoscaler.go]: addressed

File: pkg/e2e/autoscaler/autoscaler.go [hunk 8]
```diff
@@ -143,55 +190,18 @@ var _ = g.Describe("[Feature:Machines] Autoscaler should", func() {
...
-				MinReplicas: 1,
-				ScaleTargetRef: caov1alpha1.CrossVersionObjectReference{
-					Name:       targetMachineSet.Name,
-					Kind:       "MachineSet",
-					APIVersion: "machine.openshift.io/v1beta1",
-				},
-			},
-		}
+		clusterAutoscaler := clusterAutoscalerResource()
+		machineAutoscaler := machineAutoscalerResource(&targetMachineSet, 1, *initialNumberOfReplicas+1)
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/e2e/autoscaler/autoscaler.go]: addressed

File: pkg/e2e/autoscaler/autoscaler.go (hunk 1 above)
Comment (ingvagabund) [reply pkg/e2e/autoscaler/autoscaler.go]: done

File: pkg/e2e/autoscaler/autoscaler.go (hunk 4 above)
Comment (frobware) [reply pkg/e2e/autoscaler/autoscaler.go]: I think there's a lot to be said in having it scoped locally if it is only used locally, so +1 from me.

File: pkg/e2e/autoscaler/autoscaler.go [hunk 9]
```diff
@@ -348,8 +349,8 @@ var _ = g.Describe("[Feature:Machines] Autoscaler should", func() {
 				glog.Errorf("error querying api for machineSet object: %v, retrying...", err)
 				return false, nil
 			}
-			glog.Infof("Initial number of replicas: %d. Current number of replicas: %d", *initialNumberOfReplicas, *ms.Spec.Replicas)
-			if *ms.Spec.Replicas > *initialNumberOfReplicas {
+			glog.Infof("Initial number of replicas: %d. Current number of replicas: %d", initialNumberOfReplicas, *ms.Spec.Replicas)
+			if *ms.Spec.Replicas > initialNumberOfReplicas {
```
Comment (frobware) [inline pkg/e2e/autoscaler/autoscaler.go]: We should use deref here too.

File: pkg/e2e/autoscaler/autoscaler.go [hunk 10]
```diff
@@ -362,7 +363,7 @@ var _ = g.Describe("[Feature:Machines] Autoscaler should", func() {
 			scaledNodesLen := int32(len(scaledNodes.Items))
 			glog.Infof("Current number of replicas: %d. Current number of nodes: %d", *ms.Spec.Replicas, scaledNodesLen)
 			// get all linked nodes (so we can wait later on their deletion)
-			return scaledNodesLen <= *ms.Spec.Replicas && scaledNodesLen <= *initialNumberOfReplicas, nil
+			return scaledNodesLen <= *ms.Spec.Replicas && scaledNodesLen <= initialNumberOfReplicas, nil
```
Comment (frobware) [inline pkg/e2e/autoscaler/autoscaler.go]: We should use deref here too.

File: pkg/e2e/infra/utils.go [hunk 11]
```diff
@@ -138,8 +138,8 @@ func getMachineSets(client runtimeclient.Client) ([]mapiv1beta1.MachineSet, erro
 	return machineSetList.Items, nil
 }
 
-// getMachineList returns a MachineList object
-func getMachineList(client runtimeclient.Client) (*mapiv1beta1.MachineList, error) {
+// getMachines returns the list of machines or an error
+func getMachines(client runtimeclient.Client) ([]mapiv1beta1.Machine, error) {
 	machineList := mapiv1beta1.MachineList{}
```
Comment (enxebre) [inline pkg/e2e/infra/utils.go]: why removing `getMachineList` instead of reusing it?

Comment (frobware) [reply pkg/e2e/infra/utils.go]: I read this as the same thing, though instead of returning the `List` it returns a slice - the latter seems better IMO. And I'm sure I've requested similar changes in other PRs this week which do the same thing.

File: pkg/e2e/infra/utils.go [hunk 12]
```diff
@@ -332,12 +314,12 @@ func getScaleClient() (scale.ScalesGetter, error) {
 
 // nodesSnapShotLogs logs the state of all the nodes in the cluster
 func nodesSnapShotLogs(client runtimeclient.Client) error {
-	nodes, err := getNodes(client)
-	if err != nil {
-		return fmt.Errorf("error calling getNodes: %v", err)
+	nodeList := corev1.NodeList{}
```
Comment (enxebre) [inline pkg/e2e/infra/utils.go]: why this change, we are consistently trying to use get*s abstraction rather than list

File: pkg/e2e/infra/utils.go (hunk 11 above)
Comment (enxebre) [reply pkg/e2e/infra/utils.go]: based on the changes afterwards we probably want to drop this commit https://github.com/openshift/cluster-api-actuator-pkg/pull/49/commits/82456b63439e03c76fbc48b8a612fae3e24bbbb7?

File: pkg/e2e/infra/infra.go [hunk 13]
```diff
@@ -144,10 +145,14 @@ var _ = g.Describe("[Feature:Machines] Managed cluster should", func() {
...
-		machineSets, err := getMachineSets(client)
+		machineSetList := &mapiv1beta1.MachineSetList{}
+		err = client.List(
+			context.TODO(),
+			(&runtimeclient.ListOptions{}).InNamespace(e2e.TestContext.MachineApiNamespace),
+			machineSetList)
 		o.Expect(err).NotTo(o.HaveOccurred())
-		o.Expect(len(machineSets)).To(o.BeNumerically(">", 2))
-		machineSet := machineSets[0]
+		o.Expect(len(machineSetList.Items)).To(o.BeNumerically(">", 2))
```
Comment (enxebre) [inline pkg/e2e/infra/infra.go]: why are we removing getMachineSets?

File: pkg/e2e/infra/utils.go [hunk 14]
```diff
@@ -138,18 +138,11 @@ func getMachineFromNode(client runtimeclient.Client, node *corev1.Node) (*mapiv1
...
-	if err := wait.PollImmediate(1*time.Second, time.Minute, func() (bool, error) {
-		if err := client.Get(context.TODO(), key, &machine); err != nil {
-			glog.Errorf("Error querying api for nodeList object: %v, retrying...", err)
-			return false, nil
-		}
-		return true, nil
-	}); err != nil {
-		glog.Errorf("Error calling getMachineFromNode: %v", err)
-		return nil, err
+	if err := client.Get(context.TODO(), runtimeclient.ObjectKey{Namespace: namespace, Name: machineName}, &machine); err != nil {
```
Comment (enxebre) [inline pkg/e2e/infra/utils.go]: don't we want to keep this short retry for transient network issues?

File: pkg/e2e/infra/infra.go (hunk 13 above)
Comment (ingvagabund) [reply pkg/e2e/infra/infra.go]: Put back

File: pkg/e2e/infra/utils.go (hunk 12 above)
Comment (ingvagabund) [reply pkg/e2e/infra/utils.go]: Put back

File: pkg/e2e/autoscaler/autoscaler.go (hunk 6 above)
Comment (ingvagabund) [reply pkg/e2e/autoscaler/autoscaler.go]: Done

File: pkg/e2e/autoscaler/autoscaler.go (hunk 9 above)
Comment (ingvagabund) [reply pkg/e2e/autoscaler/autoscaler.go]: Done

File: pkg/e2e/autoscaler/autoscaler.go (hunk 10 above)
Comment (ingvagabund) [reply pkg/e2e/autoscaler/autoscaler.go]: Done

File: pkg/e2e/autoscaler/autoscaler.go (hunk 4 above)
Comment (enxebre) [reply pkg/e2e/autoscaler/autoscaler.go]: I was suggesting pkg/e2e/autoscaler/utils.go so no details are exposed on the ginkgo file, so it's smaller and just reads user stories. Not a blocker for me though

**Comment by ingvagabund** [comment]
This obviously is not right:
```
@@ -162,14 +165,16 @@ var _ = g.Describe("[Feature:Machines] Managed cluster should", func() {
 		o.Expect(err).NotTo(o.HaveOccurred())
 
 		g.By(fmt.Sprintf("waiting for cluster to grow %d nodes. Size should be %d", clusterGrowth, intermediateClusterSize))
-		waitForClusterSizeToBeHealthy(intermediateClusterSize)
+		err = waitForClusterSizeToBeHealthy(client, initialClusterSize)
+		o.Expect(err).NotTo(o.HaveOccurred())
 
 		g.By(fmt.Sprintf("scaling in %q machineSet to %d replicas", machineSet.Name, scaleIn))
 		err = scaleMachineSet(machineSet.Name, scaleIn)
 		o.Expect(err).NotTo(o.HaveOccurred())
 
 		g.By(fmt.Sprintf("waiting for cluster to decrease %d nodes. Final size should be %d nodes", clusterDecrease, finalClusterSize))
-		waitForClusterSizeToBeHealthy(finalClusterSize)
+		err = waitForClusterSizeToBeHealthy(client, initialClusterSize)
+		o.Expect(err).NotTo(o.HaveOccurred())
```

Just changed back to `intermediateClusterSize` and `finalClusterSize`.
